<a href="https://colab.research.google.com/github/luizguilhermembp/Programa-o-funcional-fastapi/blob/main/RegressaoLogistica_ML_LuizGuilherme.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
import pandas as pd
import numpy as np
import matplotlib as plt

In [2]:
df = pd.read_csv('titanic_v2.csv', sep=',', decimal=',')
df

Survived  Pclass                                               Name  \
0           0       3                             Mr. Owen Harris Braund   
1           1       1  Mrs. John Bradley (Florence Briggs Thayer) Cum...   
2           1       3                              Miss. Laina Heikkinen   
3           1       1        Mrs. Jacques Heath (Lily May Peel) Futrelle   
4           0       3                            Mr. William Henry Allen   
..        ...     ...                                                ...   
882         0       2                               Rev. Juozas Montvila   
883         1       1                        Miss. Margaret Edith Graham   
884         0       3                     Miss. Catherine Helen Johnston   
885         1       1                               Mr. Karl Howell Behr   
886         0       3                                 Mr. Patrick Dooley   

        Sex Age  Siblings/Spouses Aboard  Parents/Children Aboard     Fare  
0      male  22                        1                        0     7.25  
1    female  38                        1                        0  71.2833  
2    female  26                        0                        0    7.925  
3    female  35                        1                        0     53.1  
4      male  35                        0                        0     8.05  
..      ...  ..                      ...                      ...      ...  
882    male  27                        0                        0       13  
883  female  19                        0                        0       30  
884  female   7                        1                        2    23.45  
885    male  26                        0                        0       30  
886    male  32                        0                        0     7.75  

[887 rows x 8 columns]

In [3]:
df.apply(lambda x: sum(x.isnull()), axis=0)

Survived                   0
Pclass                     0
Name                       0
Sex                        0
Age                        0
Siblings/Spouses Aboard    0
Parents/Children Aboard    0
Fare                       0
dtype: int64

In [4]:
from sklearn.preprocessing import LabelEncoder
var_mod =['Sex']

le = LabelEncoder()

for i in var_mod:
  df[i] = le.fit_transform(df[i])

df


Survived  Pclass                                               Name  Sex  \
0           0       3                             Mr. Owen Harris Braund    1   
1           1       1  Mrs. John Bradley (Florence Briggs Thayer) Cum...    0   
2           1       3                              Miss. Laina Heikkinen    0   
3           1       1        Mrs. Jacques Heath (Lily May Peel) Futrelle    0   
4           0       3                            Mr. William Henry Allen    1   
..        ...     ...                                                ...  ...   
882         0       2                               Rev. Juozas Montvila    1   
883         1       1                        Miss. Margaret Edith Graham    0   
884         0       3                     Miss. Catherine Helen Johnston    0   
885         1       1                               Mr. Karl Howell Behr    1   
886         0       3                                 Mr. Patrick Dooley    1   

    Age  Siblings/Spouses Aboard  Parents/Children Aboard     Fare  
0    22                        1                        0     7.25  
1    38                        1                        0  71.2833  
2    26                        0                        0    7.925  
3    35                        1                        0     53.1  
4    35                        0                        0     8.05  
..   ..                      ...                      ...      ...  
882  27                        0                        0       13  
883  19                        0                        0       30  
884   7                        1                        2    23.45  
885  26                        0                        0       30  
886  32                        0                        0     7.75  

[887 rows x 8 columns]

In [5]:
df_x = df.drop(axis=1, columns=['Survived', 'Name'])
df_x

Pclass  Sex Age  Siblings/Spouses Aboard  Parents/Children Aboard  \
0         3    1  22                        1                        0   
1         1    0  38                        1                        0   
2         3    0  26                        0                        0   
3         1    0  35                        1                        0   
4         3    1  35                        0                        0   
..      ...  ...  ..                      ...                      ...   
882       2    1  27                        0                        0   
883       1    0  19                        0                        0   
884       3    0   7                        1                        2   
885       1    1  26                        0                        0   
886       3    1  32                        0                        0   

        Fare  
0       7.25  
1    71.2833  
2      7.925  
3       53.1  
4       8.05  
..       ...  
882       13  
883       30  
884    23.45  
885       30  
886     7.75  

[887 rows x 6 columns]

In [6]:
df_y = df[['Survived']]

In [7]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(df_x, df_y, test_size=0.3, random_state=42)

In [8]:
print(X_train.shape)
print(X_test.shape)
print(y_train.shape)
print(y_test.shape)

(620, 6)
(267, 6)
(620, 1)
(267, 1)


nao tem over

In [17]:
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score, confusion_matrix

model = LogisticRegression().fit(X_train, y_train)

/usr/local/lib/python3.10/dist-packages/sklearn/utils/validation.py:1183: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


In [14]:
predict = model.predict(X_train)

print(accuracy_score(y_train, predict))
print(confusion_matrix(y_train, predict))

0.8209677419354838
[[333  46]
 [ 65 176]]


**Acurácia: 82.10%**

A acurácia no conjunto de treinamento é bastante alta, o que pode sugerir que o modelo está se ajustando bem aos dados de treinamento.

In [15]:
predict = model.predict(X_test)

print(accuracy_score(y_test, predict))
print(confusion_matrix(y_test, predict))

0.7715355805243446
[[148  18]
 [ 43  58]]


**Acurácia: 77.15%**

A acurácia no conjunto de teste é um pouco menor do que no conjunto de treinamento, o que é esperado. Isso sugere que o modelo pode estar generalizando bem, mas não tão bem quanto no treinamento.

**CONCLUSÃO**

**Desempenho do Modelo:**

*  **Treinamento**: O modelo tem uma alta acurácia de 82.10% no conjunto de treinamento. Isso indica que o modelo é bastante eficaz para classificar os dados com base nas características fornecidas.
*   **Teste**: A acurácia no conjunto de teste é de 77.15%, o que é um pouco menor, mas ainda é uma boa performance. A diferença entre a acurácia de treinamento e teste pode ser um indicativo de que o modelo não está excessivamente sobreajustado.

**Overfitting e Underfitting:**

* **Overfitting**: A diferença entre a acurácia do treinamento e do teste (4.95%) sugere uma leve tendência para overfitting. O modelo está performando melhor no conjunto de treinamento do que no conjunto de teste, o que pode indicar que ele está aprendendo muito bem as particularidades dos dados de treinamento, mas não generaliza perfeitamente para novos dados.
* **Underfitting**: Não há evidências claras de underfitting, pois a acurácia no conjunto de teste é relativamente boa. O modelo parece ser capaz de capturar padrões úteis nos dados.

